subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/StoreSnapshotsStepTest.groovy

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps
+
+import com.google.common.collect.ImmutableList
+import com.google.common.collect.ImmutableSortedMap
+import org.gradle.caching.internal.origin.OriginMetadata
+import org.gradle.internal.Try
+import org.gradle.internal.execution.CurrentSnapshotResult
+import org.gradle.internal.execution.ExecutionOutcome
+import org.gradle.internal.execution.IncrementalContext
+import org.gradle.internal.execution.history.AfterPreviousExecutionState
+import org.gradle.internal.execution.history.BeforeExecutionState
+import org.gradle.internal.execution.history.ExecutionHistoryStore
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint
+import org.gradle.internal.hash.HashCode
+import org.gradle.internal.snapshot.impl.ImplementationSnapshot
+
+class StoreSnapshotsStepTest extends StepSpec implements FingerprinterFixture {
```

Comment:
I am not sure if the unit we are testing should be each step. The other steps are changing the context accordingly and everything needs to work together so that the interactions we mock here hold true.

We had (and still have) these kind of unit tests for the task executers. I think those never caught a problem and most times made refactoring things harder since they where so focused on the one step.

Reply:
I disagree. I think these are very useful tests. The `output snapshots are stored after failed execution when there's no previous state available` both describes the design and validates its implementation. I find tests like `deletes previous output when sourceFiles are empty`, and used them in debugging problems quite a few times. I agree that `CatchExceptionStep` is not the one in the biggest need of unit testing, but the others I think are definitely useful.